### ISSUE TITLE
Ingredient & Recipes

### ISSUE THREAD

**Issue by Ylanv** [PlateSwipe/PlateSwipe#253]
# Problem

Currently, we store the ingredient names directly in a recipe instead of their corresponding **UIDs**. This design creates challenges when trying to access ingredients within the Firestore Collection at scale. Accessing ingredients by name is not feasible for large-scale operations and limits future functionality.

---

# Proposed Solutions

### **Option 1: Keep the current implementation**

By keeping the current implementation, the following constraints and benefits arise:

#### **Constraints**
1. **Feature Limitations**:  
   - We will be unable to link recipes with the fridge inventory effectively (e.g., displaying the number of ingredients available in the fridge for a recipe).  #7 
   - Navigating to an ingredient description page from a recipe will remain unimplemented and likely unfeasible. #27 

2. **Scalability Concerns**:  
   Manual mapping of all **738** ingredients in the Recipes database to UIDs in the Firestore Collection would be avoided. However, this leaves a potential mismatch between recipe ingredients and product data from OpenFoodFacts (e.g., a recipe listing "chicken" without specifying a particular brand or product source).

#### **Benefits**
- **Reduced Workload**: No need to manually edit all existing ingredients in the database or translate them to UIDs.
- **Offline Mode Simplification**: Ingredient databases for fridge inventory and recipes can remain separate, simplifying offline mode operations.

#### Example 

![image](https://github.com/user-attachments/assets/d40fa0e9-708c-40ab-9cc6-103252064e7c)

For this recipe we should map Potatoes to a the UID of potatoes inside our firestore database, but OpenFoodFact has real item meaning that when you search for potatoes you got the following : 

![image](https://github.com/user-attachments/assets/801053d3-8228-4876-aa86-c9c8000419a1)

which is absolutely not accurate.

---

### **Option 2: Transition to UID-based implementation**

This solution involves manually updating the 738 existing ingredients in the Recipes database to their corresponding UIDs in the Firestore Collection and modifying the `ListIngredientScreen` logic.

#### **Required Changes**
The following line in `ListIngredientScreen` (line 195):
```kotlin
if (!quantity.isNullOrEmpty()) {
    createRecipeViewModel.addIngredientAndMeasurement(
              ingredient.name, quantity.toString())
}
```
would need to be updated to:
```kotlin
if (!quantity.isNullOrEmpty()) {
    createRecipeViewModel.addIngredientAndMeasurement(
              ingredient.uid, quantity.toString())
}
```

#### **Benefits**
- **Enhanced Features**:  
   - Link recipes with fridge inventory, allowing us to display the number of available ingredients for a recipe.  
   - Enable navigation to ingredient description pages (future feature).  

- **Unified Databases**: Recipes and fridge inventories will share the same UID-based ingredient structure, simplifying database management and feature integration.

#### **Challenges**
- **Manual Updates**: Manually updating 738 ingredients in the Recipes database to UIDs is labor-intensive.  
- **Data Ambiguity**: Certain ingredients may not map directly to products in OpenFoodFacts (e.g., generic references like "chicken" without brand or origin).

---

# **Request for Input**

I would appreciate input on the following:
1. Which approach aligns best with the project's long-term goals?  
2. Are there alternative solutions that might mitigate the challenges outlined above?  
3. If transitioning to UIDs, should we automate the mapping process to minimize manual work?

**Comment by RagdollGoDown** [comment]
Do you think it would be possible to map these to a OpenFoodFacts category instead? That way it would be easier to link different brand ingredients for "chicken" or "milk". It would also mean that to know if you've got an ingredient in your fridge, you can just check if they share a category.

I looked at the effectiveness of the categories and there is varying degrees of success:

Ketchup category
![image](https://github.com/user-attachments/assets/ac19a23a-7f58-49d1-9db6-8b84489e2f12)

Potatoe
![image](https://github.com/user-attachments/assets/6f7ffd8a-f962-4c2c-90fc-d4762e8808c4)

**Comment by Ylanv** [comment]
First of all thanks for your comment! 

If I understand correctly, the goal is to leverage the categories field inside OpenFoodFact to remove the branding of item. To do so when a recipe is created, we would store a list containing the categories associated with each ingredient, structured as follows: 

```kotlin
data class Recipe(
    ...
    val ingredientsAndMeasurements: List<Pair<String, String>>,
    ...
)
```

Here, the `Pair` represents `<Category, Quantity>`. This approach could facilitate matching the categories of items in the fridge with those required for a recipe. However, there are several challenges to consider:

1. **Multiple Categories per Ingredient**  
   Each ingredient can belong to multiple categories. For instance, scanning a pack of eggs purchased from a local store with the following query:  
   ```bash
   curl -X GET "https://world.openfoodfacts.net/api/v2/search?categories_tags=chicken" \
   -H 'accept: application/json' 
   ```  
   yields the following response:  
   ```json
   {
     "code": "4056489674467",
     "product": {
       "categories_tags": [
         "en:farming-products",
         "en:eggs",
         "en:chicken-eggs"
       ]
     },
     "status": 1,
     "status_verbose": "product found"
   }
   ```  
   As we can see, there are three categories returned, but only two (`"eggs"` and `"chicken-eggs"`) are relevant for fridge-recipe matching. Categories like `"farming-products"` are too generic. Therefore, we would need a method to filter and prioritize meaningful categories with sufficient accuracy, which could be a time-consuming and costly process.

2. **Firestore Querying Limitations**  
   This approach doesn’t address the issue of directly searching by fields in our Firestore collection. Additionally, if I understand correctly, we currently have a limitation of 5 searches per hour, which is not adequate for building a reliable feature around it.  
   From the [Firestore documentation](https://firebase.google.com/docs/firestore/query-data/queries), it seems we could potentially do something like the following:  
   ```kotlin
   // Create a reference to the ingredient collection
   val ingredientDB = db.collection("ingredient")

   // Create a query against the collection.
   val query = ingredientDB.whereEqualTo("Categories", "Eggs")
   ```  
   However, I might be missing some context here.

---

In conclusion, while using categories could be an effective way to abstract away branding details of items, this approach introduces significant uncertainty. Given that we are in the final stages of the project, I am hesitant to pursue this path due to the risks and additional exploration required. What are your thoughts?

**Comment by RagdollGoDown** [comment]
First, I think checking the actually quantity of the ingredients would be too much work as we would be comparing numbers in strings with units and such. 

For the second challenge, it was 5 per minute and only when searching open food facts.

Here's a description of my idea using only the categories.

We store them as so:
```kotlin
data class Ingredient(
    ...
    val categories: List<String>
)
```
When getting ingredients, no matter the repository, we can get the categories as openfoodfacts has them and we can simply save them in firestore once we got them.

In our userviewmodel, we can have an attribute `ContainedCategories` or something like that with all the categories. Then we would implement a search with filters function in the firestore recipe db (This shouldn't be too hard to do). There are filters linked to arrays in firestore. You can do Filter.whereArrayContainsAny("categories", userviewmodel.ContainedCategories)

Meanwhile in our recipes, we would need a way to keep the contained categories. Like a simple categories attribute. 
This also brings up an issue I have with our recipes; they only have specific ingredients. We aren't using ingredients but products in our recipes. It might be better to just have generic ingredients like Pepper, Milk, Apples, etc. Although this might be a bit complicated, it would allow us to not only filter by the categories of the fridge's ingredients, but simplify the creation of recipes. All people would have to do is go through these generic ones. I know we're getting close to the end so it might be too late to start developing something like this. What are your thoughts?

**Comment by Ylanv** [comment]
Let me try to recap to ensure we're on the same page. When a user creates a recipe, they can add ingredients to it. These ingredients are fetched from OpenFoodFacts, but instead of being general ingredients, they are specific products, as you mentioned. Currently, we store the name of these ingredients inside a list in the `Recipe` data class as follows:

```kotlin
data class Recipe(
    ...
    val ingredientsAndMeasurements: List<Pair<String, String>>,
    ...
) 
```

Here, the `Pair<String, String>` represents the **name of the product** and the **quantity**. For example, if the user scans the barcode `4056489674467`, it will result in:  (Using Agribalyse TAG)

```kotlin
Pair<"Egg, raw", "100 g">
```

If the user scans another brand of eggs, such as barcode `3564700814628`, the name is still:

```JSON
"name_en": "Egg, raw"
```

This is ideal because it allows us to match a product in the fridge with an ingredient in a recipe using the name.

---

### Challenges Identified

1. **Naming Consistency Between OpenFoodFacts and MealDB**  
   While OpenFoodFacts uses a relatively consistent naming convention (e.g., `"name_en": "Egg, raw"`), recipes from MealDB do not use the same names. For example, MealDB might refer to the ingredient as `"Eggs"`, while OpenFoodFacts uses `"Egg, raw"`. This inconsistency makes matching difficult, but im willing to do the migration manually if needed.

2. **Field Selection for Ingredient Names**  
   Currently, we are storing the `product_name` field from OpenFoodFacts, but this might not be the best choice. OpenFoodFacts provides multiple fields related to ingredient names, including:
   - `"product_name"`: The name of the specific product (e.g., `"Oeufs frais plein air Oméga3 x 12"`) — too specific for our use case.
   - `"generic_name"`: The regulated, legal name, often empty.
   - `"agribalyse"`: Information from a separate database (e.g., `"name_en": "Egg, raw"` and `"name_fr": "Oeuf, cru"`). This seems like a more general and coherent option.

   For example, for the barcode `3564700814628`, the JSON includes the following fields:
   ```JSON
   "product_name": "Oeufs frais plein air Oméga3 x 12",
   "agribalyse": {
       "name_en": "Egg, raw",
       "name_fr": "Oeuf, cru"
   },
   "generic_name": "",
   "generic_name_fr": ""
   ```
   Based on the documentation:
   - `"generic_name"` is the legal name but is often unavailable.
   - `"product_name"` is too specific.
   - `"agribalyse"` seems promising, but we need to ensure that it is widely available across OpenFoodFacts entries.

---

### Proposed Approach

1. **Generalization of Ingredient Names**  
   The core idea is to store the **generalized name** of the ingredient in the recipe instead of the product’s UID. This simplifies offline functionality since we can have a separate database for the fridge and recipes. For barcode scanning and searching, this also helps match ingredients more reliably.

2. **Exploring `agribalyse` as a Candidate**  
   The `"agribalyse"` field appears to offer a more generic name for ingredients and could be a suitable option. However, we should confirm its availability across a significant portion of OpenFoodFacts products to avoid gaps.

3. **Avoid Using Categories for Naming**  
   Using categories to name products introduces too much ambiguity and is likely not the best approach. Instead, focusing on a specific field like `"agribalyse"` might provide the clarity we need.

---

### Conclusion and Next Steps

The overall idea is to rely on a consistent, generalized naming convention for ingredients in recipes. This allows us to:
- Match ingredients between the fridge and recipes using names.
- Simplify offline functionality with separate databases for recipes and the virtual fridge.

The key challenge remains identifying the most reliable field for naming ingredients. I suggest we explore the `"agribalyse"` field further to determine its feasibility. 

What are your thoughts on this approach? Do you think focusing on `"agribalyse"` could address the generalization issue effectively, or do you have other ideas in mind ? Let me know!

Please note that this entire analysis is based on a limited sample of items and may not be fully representative of all the items.

**Comment by RagdollGoDown** [comment]
Unfortunately, Agribalyse actually refers to a database, but it's limited to 2500 products so most of the ingredients from open food facts won't have it. Even those that do, often reference their agribalyse code instead.

We should definitely talk about this next meeting

**Comment by Ylanv** [comment]
**Subject:** Enhancing Ingredient Management for Recipe Creation

**Introduction:**  
After discussing with my colleague, @RagdollGoDown, and considering his valuable input, we identified an improved approach to how users can add ingredients to recipes during creation. Our goal is to address the limitations of the current implementation and improve functionality while making the process more user-friendly and efficient.

---

### **Current Implementation and Its Challenges**  
Currently, users can add ingredients by either:  
1. **Scanning a product**, or  
2. **Using the search bar** to find ingredients.  

**Problem with Product Name Dependency:**  
Both methods fetch ingredients based on the product_name field from OpenFoodFacts, resulting in non-generic names. For example:  
- Scanning two different brands of egg packets produces distinct values for `product_name`.  

This makes it difficult to standardize ingredients across users. We want a system that maps these product-specific names to **generic ingredient names** (e.g., "Egg").

**Limitation of the Search Bar:**  
The search bar further complicates this issue because OpenFoodFacts restricts ingredient name queries to **five searches per minute**.

---

### **Proposed Solution**  
We will leverage **OpenFoodFacts categories** to address these issues. The improved workflow is as follows:  
1. Users will now select ingredients by browsing or searching for **categories** instead of product names.  
2. A recipe will store ingredients in the following structure:  
   ```kotlin
   data class Recipe(
       ...
       val ingredientsAndMeasurements: List<Pair<String, String>>,
       ...
   )
   ```
   Here, `Pair<Category, Quantity>` ensures a consistent, generic representation of ingredients.

---

### **How This Solves the Problem**  
Each ingredient in OpenFoodFacts contains a list of categories. By matching a recipe's required categories with those in the user's fridge, we achieve generic ingredient mapping.  

**Example:**  
- **Recipe Ingredients:**  
   ```kotlin
   List(Pair("Eggs", "6"), Pair("Butter", "100g"), Pair("Sugar", "150g"))
   ```  
- **Fridge Ingredient Example:**  
   ```kotlin
   Ingredient(
       categories = List("Produit de la ferme", "Eggs", "Raw", ...)
   )
   ```  

Since the category list for the ingredient includes "Eggs," we can match this to the recipe's requirement and confirm the user has eggs in their fridge.

---

### **Additional Features and Considerations**  

1. **Searching for Ingredients by Categories:**  
   - Users will search within a list of categories retrieved from [OpenFoodFacts](https://world.openfoodfacts.org/data).  
   - However, we encountered an issue: OpenFoodFacts has **70,962 categories**, but some essential ones, such as "Egg," are missing on the web page despite being present in scanned JSON data.  

![image](https://github.com/user-attachments/assets/26f664b3-669c-42c7-b936-e38027b817ac)


2. **Possible Solution for Missing Categories:**  
   - Download the full dataset as a `.csv` file from [OpenFoodFacts](https://world.openfoodfacts.org/data).  
   - Use a script to extract all available categories.  

3. **Storing Categories:**  
   - Categories can either be stored **locally** or in a **Firestore collection**, depending on performance and scalability requirements.

4. **Offline Mode Compatibility:**  
   - This new method does not introduce complications for offline mode since the categories can be pre-fetched and stored locally.

5. **Drawback:** 
- The scanner will not be usable inside the created recipe
---

### **Conclusion**  
By transitioning to a category-based ingredient selection system, we ensure consistency across users and overcome the limitations of product_name-based fetching. While there are challenges related to missing categories and storage, these can be addressed with pre-processing and database solutions.

**Comment by 2Fick** [comment]
Hey everyone!

I love this idea, switching to categories instead of product names is such a smart move! It’ll make things way more consistent and user-friendly. The example with "Eggs" really helped me see how it all ties together.

For the missing categories, extracting them from the full dataset sounds like a great plan. Maybe we could also set up a way to keep it updated over time? As for storage, local storage feels like a simple choice for offline mode, but Firestore could work if we need more scalability.

The scanner not working in recipes is a bit of a bummer, but maybe we could figure out a way to map scanned products to categories down the line?

Overall, I’m totally on board with this. Let me know if there’s anything I can do to help out and this is exciting stuff! 🚀

**Comment by Ylanv** [comment]
Thanks for the feedback @2Fick!

**Comment by akshire** [comment]
Thank you for sharing your detailed plan! The shift to category-based ingredient selection is a significant improvement, addressing the inconsistency caused by product name dependency. This approach ensures standardization, enhances user experience, and provides a more intuitive way to add ingredients. Using OpenFoodFacts categories to match recipe requirements with fridge contents is a practical and scalable solution.

The issue of missing categories, such as "Eggs," is a challenge, but your idea of extracting data from the full dataset is a solid fix. Adding a fallback for manual category mapping could further enhance flexibility. Offline mode compatibility is a strength, though periodic updates to the category dataset may be needed to maintain accuracy.

The scanner's reduced functionality is a drawback, but exploring a hybrid system that maps scanned items to categories could mitigate this. Additionally, optimizing category storage to improve performance on lower-end devices is essential.

Overall, this is a well-thought-out solution that addresses key issues and opens the door for exciting future features like fridge-based recipe suggestions. Let me know if you need any support in implementation or testing!

**Comment by AdriB02** [comment]
This is a great idea because it solves the current problems and makes the system easier to use. Switching to a category-based ingredient selection means users will see simple, generic names like "Egg" instead of brand-specific ones. This change helps match recipe needs with what users have in their fridge, making the process smoother. It also avoids issues with OpenFoodFacts’ search limits, making the system more reliable.

Storing categories locally or in a database is a smart way to handle missing categories and allows the app to work offline. While the scanner won’t work in this new setup, the benefits of easier, more consistent ingredient selection make up for it. This solution will make recipe creation faster, clearer, and more user-friendly.
